### ISSUE TITLE
[API Proposal]: Creating new TypeName with given AssemblyNameInfo

### ISSUE THREAD

**Issue by adamsitnik** [dotnet/runtime#102263]
### Background and motivation

As we have started using the new `TypeName` APIs all over the place, I've realized that it would be very nice to have the ability to create a new `TypeName` instance with given `AssemblyNameInfo` without the need of re-parsing the [concatenated input](https://github.com/dotnet/winforms/blob/31176a4d203711e766b8b587414d04c27d8cc173/src/System.Windows.Forms/src/System/Resources/ResXSerializationBinder.cs#L54-L61) again.

For example, in BinaryFormatter the type and library names are provided separately (https://github.com/dotnet/runtime/issues/102014#issuecomment-2112683887), I would like to join them without re-parsing the whole thing again.

### API Proposal

```csharp
namespace System.Reflection.Metadata;

public sealed class TypeName
{
    /// <summary>
    /// Creates a new TypeName with provided assembly name.
    /// </summary>
    public TypeName With(AssemblyNameInfo newAssemblyNameInfo);
}
```

### API Usage

```csharp
TypeName WithoutAssemblyVersion(TypeName typeName) 
    => typeName.With(
        new AssemblyNameInfo(
            typeName.AssemblyName.Name,
            version: null,
            typeName.AssemblyName.CultureName,
            typeName.AssemblyName.Flags,
            typeName.AssemblyName.PublicKeyOrToken));
```

### Alternative Designs

The name could be changed to sth that indicates that it creates and returns a new instance.

### Risks

Some users may miss the fact that the API creates a new instance of `TypeName` rather than mutating the current instance.

**Comment by jkotas** [comment]
Should we rather address the more general problem of creating TypeName of any shape through constructors or factory methods?

It happens that the AssemblyName case showed up with the binary formatter. I think it is a tell-take of a general need for creating arbitrary TypeNames.

**Comment by adamsitnik** [comment]
> Should we rather address the more general problem of creating TypeName of any shape through constructors or factory methods?

All `TypeName` ctors that are going to accept strings of any kind are going to need to be fully validated, which is almost like reparsing them. And this is something that I would like to avoid with the provided API.

I am not saying no, I just don't have a clear vision of how these APIs would look like.

**Comment by jkotas** [comment]
> All TypeName ctors that are going to accept strings of any kind are going to need to be fully validate

Why do the strings need to be validated? Type names can be anything.

**Comment by jkotas** [comment]
Ah, it is because the APIs returned escaped names. The escaping validation should be very cheap, much cheaper than parsing.

The more I see the issues with the name escaping, the more I think that it was a bad design decision to make the Name and FullName properties to match reflection behavior. Should we revisit that?

The APIs can look like this:
```
static TypeName MakeSimpleTypeName(string fullName, TypeName? declaringType = null, AssemblyNameInfo? assemblyInfo = null);
static TypeName MakeArrayTypeName(TypeName elementType);
static TypeName MakeArrayTypeName(TypeName elementType, int rank);
static TypeName MakePointerTypeName(TypeName elementType);
static TypeName MakeByRefTypeName(TypeName elementType);
static TypeName MakeGenericTypeName(TypeName genericTypeDefinition, ImmutableArray<TypeName> typeArguments);
```

**Comment by jkotas** [comment]
> if (typeName.FullName == typeof(T).FullName)

Yeah, that's the conundrum when you want to use this together with reflection. It is hard to tell how common this use case is going to be. 

Anyway, I edited my suggestion at https://github.com/dotnet/runtime/issues/102263#issuecomment-2115649352 to include `Escape` method, so that there is a reasonable path when you have the raw name.

> we would not have that problem if we would simply forbid using certain characters without the ability to escape them in the first place.

It would produce non-compliant type name parser. It may be ok for BF-specific parser. I do not think it is ok for S.R.Metadata or the runtime that need compliant and backward compatible type name parsers. It would be a needless breaking change to drop support for escaped type names.

**Comment by steveharter** [comment]
Moving to v10 since we have reached the "feature complete" milestone; still in API review backlog.

**Comment by bartonjs** [comment]
* WithAssemblyName doesn't seem to have a solid scenario (outside of one case), so it's being withheld until it has a more clear use pattern with regard to composed types (e.g. generics).
* It was asked if the instance methods should be prefixed with "Make" or "Create", and it seems that "Make" matches reflection, and the static one "Create" is a patterned prefix.
* We changed CreateSimpleTypeName's `fullName` parameter to `metadataName` to avoid confusion between the unescaped value as a parameter and the escaped value as the property with the same name.
* The parameterless MakeArrayTypeName is renamed to MakeSZArrayTypeName because MakeArrayTypeName(1) and MakeArrayTypeName() are not the same.

```c#
namespace System.Reflection.Metadata;

public partial class TypeName
{      
    public TypeName MakeSZArrayTypeName();
    public TypeName MakeArrayTypeName(int rank);
    public TypeName MakePointerTypeName();
    public TypeName MakeByRefTypeName();
    public TypeName MakeGenericTypeName(ImmutableArray<TypeName> typeArguments);

    public static TypeName CreateSimpleTypeName(
        string metadataName,
        TypeName? declaringType = null,
        AssemblyNameInfo? assemblyInfo = null);
}
```

**Comment by adamsitnik** [comment]
I've adopted my initial implementation (#103713) to the approved API in https://github.com/dotnet/runtime/pull/103713/commits/9a0cd5819f8d5367e94aeb0da9d98e6fb5ea51f3.

So far, all `TypeName` instances that any library or app could somehow get were valid and escaped type names. Users could simply rely on that and the information the API provides no matter where it comes from.

The introduction of `CreateSimpleTypeName` that accepts unescaped full names introduces a possibility to create invalid `TypeName` instances.

Example: `TypeName.Parse("TypeName, MyLib")` returns a `TypeName` where `FullName=TypeName` and `AssemblyName=MyLib`.
When we pass the same input to `CreateSimpleTypeName` we don't know whether `,` was previously escaped (it's not a begining of assembly name) or not (it's a begining of an assembly name).

Same goes for escaping. If we have a name that consists of escaped escape character followed by an escaped special character like `*` and unescape it, the escape character is followed by a special character. It seems to still be escaped, but it's not.

Example: `\\\*` -> Unescape -> `\*` -> Unescape -> `*`.

In my opinion, since `static string Unescape(string name)` got approved https://github.com/dotnet/runtime/issues/101774#issuecomment-2263693421, we don't need `CreateSimpleTypeName`.

But we still need the ability to replace `AssemblyName` for simple types. That is why I propose following change:

```diff
public class TypeName
{

-   public static TypeName CreateSimpleTypeName(string metadataName, TypeName? declaringType = null, AssemblyNameInfo? assemblyInfo = null);
+   public TypeName MakeSimpleTypeName(TypeName? declaringType = null, AssemblyNameInfo? assemblyInfo = null);
}
```

That method would throw for non-simple names and prevent the users from creating invalid instances. I've implemented it in https://github.com/dotnet/runtime/pull/103713/commits/aabe51b8bfb7e2ab7d90d1672e4feb56f91315af

cc @jkotas @GrabYourPitchforks

**Comment by jkotas** [comment]
> Example: TypeName.Parse("TypeName, MyLib") returns a TypeName where FullName=TypeName and AssemblyName=MyLib.
When we pass the same input to CreateSimpleTypeName we don't know whether , was previously escaped (it's not a begining of assembly name) or not (it's a begining of an assembly name).

`CreateSimpleTypeName` was meant to take unescaped name. It is why we have opted to name the argument `metadataName` to make it clear that it is not the regular (escaped) name.

If you pass `TypeName, MyLib` to `CreateSimpleTypeName`, it should create a TypeName that returns `TypeName\, MyLib` from FullName property. It is a valid typename. There is no ambiguity.

> In my opinion, since static string Unescape(string name) got approved https://github.com/dotnet/runtime/issues/101774#issuecomment-2263693421, we don't need CreateSimpleTypeName.

`CreateSimpleTypeName` was equivalent of `Escape`. `Unescape` is not a replacement.

**Comment by adamsitnik** [comment]
@jkotas big thanks for the clarification, now it all makes sense.

Since `CreateSimpleTypeName` needs to parse the name to unescape it, I can not use it (my requirement is to create a new type name with given assembly name without reparsing).

I've finished implementing https://github.com/dotnet/runtime/pull/103713 and I need just one more method:

```diff
namespace System.Reflection.Metadata;

public partial class TypeName
{      
    public TypeName MakeSZArrayTypeName();
    public TypeName MakeArrayTypeName(int rank);
    public TypeName MakePointerTypeName();
    public TypeName MakeByRefTypeName();
    public TypeName MakeGenericTypeName(ImmutableArray<TypeName> typeArguments);
+   public TypeName MakeSimpleTypeName(AssemblyNameInfo? assemblyInfo);

    public static TypeName CreateSimpleTypeName(
        string metadataName,
        TypeName? declaringType = null,
        AssemblyNameInfo? assemblyInfo = null);
}
```

I am going to create a new issue with the proposal for it.

**Comment by jkotas** [comment]
> Since CreateSimpleTypeName needs to parse the name to unescape it, I can not use it (my requirement is to create a new type name with given assembly name without reparsing).

I do not see why escaping the type name passed into `CreateSimpleTypeName` is a problem for your BF use case. It does not create non-linear amount of work.

**Comment by adamsitnik** [comment]
> I do not see why escaping the type name passed into `CreateSimpleTypeName` is a problem for your BF use case.
> It does not create non-linear amount of work. 

I think it depends on how you treat the comparison of current char vs chars that needs to be escaped. For me, assuming that we have `n` needles (characters that need escaping) and `i` characters in the input name, it's `O(n * i)`.

**Comment by jkotas** [comment]
This is not a general-purpose escaping with arbitrary escape characters. The number of characters that need escaping is very small (7 to be exact) and set in stone. So it is `O(7 * i)`.

**Comment by teo-tsirpanis** [comment]
@adamsitnik is there anything left to do here? Can we close this issue?

**Comment by adamsitnik** [comment]
> @adamsitnik is there anything left to do here? Can we close this issue?

The `CreateSimpleTypeName` method has not been implemented yet.

**Comment by jkotas** [comment]
> The CreateSimpleTypeName method has not been implemented yet.

That's correct. On the other hand, I am not aware of any place that can use the API of this shape currently, so we may want to let this issue linger until this need materializes.

There are a few places that do raw type name escaping (e.g. https://github.com/dotnet/runtime/blob/2a1b35e8afe469e4db01de7ba96b9133acc55d38/src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/General/Helpers.cs#L103 ), but this API is not of the right shape to replace them.
